### ISSUE TITLE
Windows PATH variable pieces can be in quotes

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7450]
Migrated from [rt.perl.org#30966](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=30966) (status was 'resolved')

Searchable as RT30966$

**Comment by p5pRT** [comment]
### From warren.dodge@tektronix.com

#### Created by warrend@tek.com




I just encountered a problem in a script that parses 
the PATH variable and checks to make sure each entry 
has a directory\.

It turns out that some program entered something like this
as a PATH entry\.   "C&#8203;:\\PATH\\BIN"

The quotes were included\. I tested this at a CMD prompt 
and indeed it does work as expected\. The quotes are shown in a 
set path  command\.

It seems the \-d operator in perl doesn't handle this properly

if \( \-d "\\"C&#8203;:/temp\\"" \)
\{
&nbsp;  print "OK1\\n";
\}
else
\{
&nbsp;  print "BAD1\\n";
\}
if \( \-d "C&#8203;:/temp" \)
\{
&nbsp;  print "OK2\\n";
\}
else
\{
&nbsp;  print "BAD2\\n";
\}

So I wanted to let you know so perhaps the low level stuff 
could be enhanced\.

<details><summary>Perl Info</summary>

```

Flags:
    category=utilities
    severity=low

Site configuration information for perl v5.8.4:

Configured by ActiveState at Tue Jun  1 11:52:09 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl584\lib\CORE"  -machine:x86'
    libpth=C:\PROGRA~1\MICROS~3\VC98\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl584\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    22751 Update to Test.pm 1.25
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.4:
    C:/Perl584/lib
    C:/Perl584/site/lib
    .


Environment for perl v5.8.4:
    HOME=U:\
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Perl584\bin\;j:\apps\perl561\bin;c:\oracle\ora92\bin;C:\Program Files\Oracle\jre\1.3.1\bin;C:\Program Files\Oracle\jre\1.1.8\bin;C:\WINNT\System32\Wbem;c:\emacs-21.2\bin;C:\Program Files\Microsoft SMS Installer\Installer Step-up Utility\Apps;C:\Program Files\Microsoft SMS Installer\Installer Step-up Utility\Apps\Extensions;c:\cygwin\USR\X11R6;c:\cygwin\USR\X11R6\BIN;C:\WINNT\system32;C:\WINNT;C:\EDPM\java\jdk1.3.1_03\jre\bin\hotspot;C:\EDPM\bin;J:\apps\edatools\psd\15.2a\tools\pcb\bin;J:\apps\edatools\psd\15.2a\tools\bin;J:\apps\edatools\psd\15.2a\tools\Capture;J:\apps\edatools\psd\15.2a\tools\fet\bin;J:\apps\edatools\psd\15.2a\tools\specctra\bin;J:\apps\edatools\psd\15.2a\tools\libutil\bin;J:\apps\edatools\psd\15.2a\tools\jre\bin;J:\apps\edatools\psd\15.2a\tools\pspice;J:\apps\edatools\fetools\psd152\bin;J:\apps\edatools\cadtools\psd152\bin;"C:\temp"
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Aug 06\, 2004 at 04&#8203;:34&#8203;:56PM \-0000\, Warren Dodge wrote&#8203;:
> I just encountered a problem in a script that parses 
> the PATH variable and checks to make sure each entry 
> has a directory\.
> 
> It turns out that some program entered something like this
> as a PATH entry\.   "C&#8203;:\\PATH\\BIN"
> 
> The quotes were included\. I tested this at a CMD prompt 
> and indeed it does work as expected\. The quotes are shown in a 
> set path  command\.
> 
> It seems the \-d operator in perl doesn't handle this properly
> 
> if \( \-d "\\"C&#8203;:/temp\\"" \)
> \{
>   print "OK1\\n";
> \}
> else
> \{
>   print "BAD1\\n";
> \}
> if \( \-d "C&#8203;:/temp" \)
> \{
>   print "OK2\\n";
> \}
> else
> \{
>   print "BAD2\\n";
> \}
> 
> So I wanted to let you know so perhaps the low level stuff 
> could be enhanced\.

Thanks for the report\.

On reflection\, I don't think the basic OS functions should be messed with
the emiulate the particular quirks associated with teh $PATH environment
variable\.

Dave\.

\-\- 
Little fly\, thy summer's play my thoughtless hand
has terminated with extreme prejudice\.
&nbsp;        \(with apologies to William Blake\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

> > It turns out that some program entered something like this
> > as a PATH entry\.   "C&#8203;:\\PATH\\BIN"
> > 
> > The quotes were included\. I tested this at a CMD prompt 
> > and indeed it does work as expected\. The quotes are shown in a 
> > set path  command\.
> > 
> > It seems the \-d operator in perl doesn't handle this properly

Personally I wouldn't expect the \-d operator to handle this correctly as it
expects a string that represents a path\, so it doesn't need quoting
semantics\.

However I would expect that File&#8203;::Spec\->path\(\) would handle it correctly and
it doesn't\.

File&#8203;::Spec&#8203;::Win32&#8203;::path\(\) should probably look like the following&#8203;:

&nbsp;  sub path \{
&nbsp;      my $pathstr=shift;
&nbsp;      $pathstr=$ENV\{path\} unless defined $pathstr;
&nbsp;    @&#8203;path=\(''\);
&nbsp;    while \($pathstr=~/\\G\(?&#8203;:\(;\)|\(\[^;"\]\+\)|"\(\[^"\]\+\)"\)/g\) \{
&nbsp;        if \($1\) \{
&nbsp;            push @&#8203;path\,'';
&nbsp;        \} else \{
&nbsp;            $path\[\-1\]\.=$2||$3;
&nbsp;        \}
&nbsp;    \}
&nbsp;    pop @&#8203;path unless length $path\[\-1\];
&nbsp;    my %seen;
&nbsp;    return map \{ $\_=$\_ eq '' ? '\.' : $\_; 
&nbsp;                   $seen\{$\_\}\+\+ ? \(\) : $\_ \} @&#8203;path;
&nbsp;  \}

In fact this is yet another good reason why F&#8203;::S\->path\(\) should optionally
take an argument\. These type of quoting rules will apply to almost any path
setting in Win32\, and even if I say it myself the above code is not trivial
enough to expect newbies to write\. Id actually go so far as to bet a lot of
win32 Perl programmers don't even know all the rules for how quotes can be
placed in path strings\.

For instance\, try the following commands&#8203;:

&nbsp;  >md "a;b"
&nbsp;  >set path=%PATH%;a";"b
&nbsp;  >copy con "a;b/weird\_test\.bat"
&nbsp;  @&#8203;echo Here\!
&nbsp;  ^Z
&nbsp;  >weird\_test\.bat

And youll see that weird\_test\.bat will be found and executed\. Since this is
true IMO F&#8203;::S\->path\(\) should ensure that in its returned values it includes
qq\(a;b\) with the quotes properly removed\.

Also\, a note to whomever wrote the existing F&#8203;::S&#8203;::Win32\->path\(\)\, on Win32
the ENV hash is case insensitive\.

Yves
Ps&#8203;: I couldn't see from my version of the docs who is responsible for
F&#8203;::S&#8203;::Win32 so Im not sure who to send this too\. Is it Ken Williams?





&nbsp;

**Comment by p5pRT** [comment]
### From warren.dodge@tektronix.com



I wanted to make sure it was understood the the quoted PATH entry was put in
there by some application install\. The test case I gave just hard coded a
quoted path to demonstrate\.

In my case I was parsing all the PATH pieces and the quoted one failed\.


As mentioned there are probably a lot of combinations that work that aren't
covered in the code\.

My thought was perhaps the \-d test could just remove a leading/trailing quote
character prior to the tests\.

I added code to protect against this condition and am happy for now\.

**Comment by p5pRT** [comment]
### From @rgs

This should be fixed by this change&#8203;:

Change 29690 on 2007/01/05 by rgs@&#8203;marais

&nbsp;        Subject&#8203;: \[PATCH\] ANSIfy the PATH environment variable on Windows
&nbsp;        From&#8203;: Jan Dubois \<jand@&#8203;ActiveState\.com>
&nbsp;        Date&#8203;: Thu\, 04 Jan 2007 12&#8203;:37&#8203;:48 \-0800
&nbsp;        Message\-ID&#8203;: \<7coqp2pme3r30qaonticuv6c6licieg4cg@&#8203;4ax\.com>

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
